server/src/main/java/org/opensearch/cluster/coordination/ClusterStateUtils.java

```diff
@@ -0,0 +1,98 @@
+/*
+ * SPDX-License-Identifier: Apache-2.0
+ *
+ * The OpenSearch Contributors require contributions made to
+ * this file be licensed under the Apache-2.0 license or a
+ * compatible open source license.
+ */
+
+package org.opensearch.cluster.coordination;
+
+import org.apache.logging.log4j.LogManager;
+import org.apache.logging.log4j.Logger;
+import org.opensearch.cluster.ClusterState;
+import org.opensearch.cluster.Diff;
+import org.opensearch.cluster.node.DiscoveryNode;
+import org.opensearch.common.bytes.BytesReference;
+import org.opensearch.common.compress.Compressor;
+import org.opensearch.common.compress.CompressorFactory;
+import org.opensearch.common.io.stream.BytesStreamOutput;
+import org.opensearch.common.io.stream.InputStreamStreamInput;
+import org.opensearch.common.io.stream.NamedWriteableAwareStreamInput;
+import org.opensearch.common.io.stream.NamedWriteableRegistry;
+import org.opensearch.common.io.stream.OutputStreamStreamOutput;
+import org.opensearch.common.io.stream.StreamInput;
+import org.opensearch.common.io.stream.StreamOutput;
+import org.opensearch.common.io.stream.Writeable;
+import org.opensearch.transport.BytesTransportRequest;
+
+import java.io.IOException;
+
+/**
+ * A helper class to utilize the compressed stream.
+ *
+ * @opensearch.internal
+ */
+public final class ClusterStateUtils {
+    private static final Logger logger = LogManager.getLogger(ClusterStateUtils.class);
+
+    /**
+     * Serialize the given cluster state or diff. It'll always use compression before writing on a newly created output
+     * stream.
+     *
+     * @param writer             Object which is going to write the content
+     * @param node               version of cluster node
+     * @param isFullClusterState flag used at receiver end to make intelligent decisions. For example, ClusterState
+     *                           assumes full state of diff of the states based on this flag.
+     * @return reference to serialized bytes
+     * @throws IOException if writing on the compressed stream is failed.
+     */
+    public static BytesReference serializeClusterState(Writeable writer, DiscoveryNode node, boolean isFullClusterState)
```

Comment:
A few things are a bit weird about the signature here. It looks like the usage of this method is either:

```
serializeClusterState(ClusterState, node, true);
```
or 

```
serializeClusterState(Diff<ClusterState>, node, false);
```

It would probably be better to use overloading and ditch the boolean, e.g.:

```
serializeClusterState(ClusterState, DiscoveryNode);
serializeClusterState(Diff<ClusterState>, DiscoveryNode);
```

That being said, the bigger issue here is that this is essentially defining a serialization schema (version number, boolean indicating whether it is full or diff, and then the content), but knowledge of how to read this schema is spread to JoinHelper and PublicationTransportHandler. It would be better to keep knowledge of this schema encapsulated in a single class. Is it possible to use a discriminated union pattern here where you have a class defined that can hold either or full cluster state or a diff and then the callers don't have to have knowledge of the serialization schema?

Reply:
That's a nice suggestion @andrross . 

>It would probably be better to use overloading and ditch the boolean, e.g.:
serializeClusterState(ClusterState, DiscoveryNode);
serializeClusterState(Diff<ClusterState>, DiscoveryNode);

The implementation is exactly same for both the cases, keeping two different method looks okay me. Internally, I'll create a single method and call it from both of them. Though a single method where we detect full/diff based on object(instanceof) is good enough. So, I'll incorporate it.

Code would look something like 

```
public static BytesReference serializeClusterState(ClusterState clusterState, DiscoveryNode node) throws IOException {
        return serializeClusterStateInternal(clusterState, node, true);
    }

public static BytesReference serializeClusterState(Diff<ClusterState> clusterStateDiff, DiscoveryNode node) throws IOException {
        return serializeClusterStateInternal(clusterStateDiff, node, false);
    }

private static BytesReference serializeClusterStateInternal(Writeable writer, DiscoveryNode node, boolean isFullClusterState)
        throws IOException {
        final BytesStreamOutput bStream = new BytesStreamOutput();
        try (StreamOutput stream = new OutputStreamStreamOutput(CompressorFactory.COMPRESSOR.threadLocalOutputStream(bStream))) {
            stream.setVersion(node.getVersion());
            stream.writeBoolean(isFullClusterState);
            writer.writeTo(stream);
        }
        final BytesReference serializedByteRef = bStream.bytes();
        logger.trace("serialized writable object for node version [{}] with size [{}]", node.getVersion(), serializedByteRef.length());
        return serializedByteRef;
    }
```

On the other suggestion :
Just so we all are on the same page, here I was trying make code changes in PublicationTransportHandler so there is no code duplication for same logic(state compression/decompression). So I'm not changing the existing code, just moved it to a common place. That's why the utility class was created. 

I hope it's okay to take this (discriminated union pattern one) change separately. I can create a new issue for this cosmetic change.

Reply:
> I was trying make code changes in PublicationTransportHandler so there is no code duplication for same logic(state compression/decompression). So I'm not changing the existing code, just moved it to a common place. That's why the utility class was created.

Thanks for the clarification. There was a nuance here that I missed. My hangup is that PublicationTransportHandler is defining a serialization schema with the following logic:

```
stream.setVersion(nodeVersion);
stream.writeBoolean(false);
diff.writeTo(stream);
```

If JoinHandler and PublicationTransportHandler are not exchanging data between each other, then I don't think they should share that _schema_. I think JoinHandler doesn't even need the boolean or the ability to write the diff, and by coupling this there's now an invalid code path in JoinHandler. Is that right?

The other downside here is that if PublicationTransportHandler needs to evolve the data that it writes (by adding a field for example), then this code will have to be untangled or JoinHandler will have to updated to handle the new field, which may or may not make sense.

Reply:
>The other downside here is that if PublicationTransportHandler needs to evolve the data that it writes (by adding a field for example), then this code will have to be untangled or JoinHandler will have to updated to handle the new field

JoinHelper and PTH are not sharing data per se, but they're doing exactly same thing. In both cases, we are sending the cluster state to other node. First we need to remember that the way these variables are written on stream, that order will never change because based on that order only all existing nodes are functional. 
In case we add a new field on PTH (a boolean or int). If it doesn't make sense for JoinHelper, JH reader doesn't need to be updated as it'll function fine. On the other hand if JH requires some new variable, if that intelligence is not required by PTH, that reader will keep functioning fine. 
If some (must know) intelligent variable is added, which can't be avoided for reading cluster state, then both reader part needs to be updated.
I'm still not able to think about any of the use cases. I made this single util class so all code related to compression, decompression, serialization, de-serialization can be put in same place, so whatever is being written on stream is available for all the readers. It's up to the reader if they want to use it or not.

I think we can wait for the time when there is a real use case evolved in this flow, if that really requires a de-coupling, I'll be happy to take it up. 
Note : I'm anyway planning to open up a new issue about this more refactoring. I'll link it here.

Reply:
The reason, code was refactored between JoinHelper and PublicationTransportHandler is both publish the cluster state and can use the same logic and optimization in future. PublicationTransportHandler has an optimization to publish diff as well but JoinHandler always publishes the full state. I don't see we publishing diff from JoinHelper so boolean is not needed.

Reply:
@andrross @shwetathareja 

Ack, I understand the concern of sending extra boolean over the stream which will not be utilized even in future. I'll keep the compression/decompression code same and separate out the code writing on the stream.

Will update the pseudo code here before pushing the actual commit, so we're all on the same page.

Reply:
There are 2 ways to achieve this separation. 

**First : Just create a new method in ClusterStateUtils class for node join which doesn't send boolean. So, we can either know about the caller from some ENUM or directly with a different method name. If we choose ENUM, existing method can then have if else etc.**

```
public static BytesReference serializeClusterStateForJoin(Writeable writer, DiscoveryNode node)
        throws IOException {
        final BytesStreamOutput bStream = new BytesStreamOutput();
        try (StreamOutput stream = new OutputStreamStreamOutput(CompressorFactory.COMPRESSOR.threadLocalOutputStream(bStream))) {
            stream.setVersion(node.getVersion());
            writer.writeTo(stream);
        }
        final BytesReference serializedByteRef = bStream.bytes();
        logger.trace("serialized writable object for node version [{}] with size [{}]", node.getVersion(), serializedByteRef.length());
        return serializedByteRef;
    }
```

**Second way would be to create separate classes for these headers which we're adding on top of actual content(cluster state).**

```
public class NodeVersionSchemaBase implements Writeable {

    private final Version nodeVersion;

    public NodeVersionSchemaBase(Version nodeVersion) {
        this.nodeVersion = nodeVersion;
    }

    public Version getNodeVersion() {
        return nodeVersion;
    }

    @Override
    public void writeTo(StreamOutput out) throws IOException {
        out.setVersion(nodeVersion);
    }
}


public class ClusterStateUpdateSchema extends NodeVersionSchemaBase implements Writeable {
    private final boolean isFullClusterState;

    public ClusterStateUpdateSchema(Version nodeVersion, boolean isFullClusterState) {
        super(nodeVersion);
        this.isFullClusterState = isFullClusterState;
    }

    @Override
    public void writeTo(StreamOutput out) throws IOException {
        super.writeTo(out);
        out.writeBoolean(isFullClusterState);
    }
}

```


Based on the schema our Util class code will be changed something like 
```
 public static BytesReference serializeClusterState(ClusterState clusterState, DiscoveryNode node) throws IOException {
        return serializeClusterStateOrDiff(clusterState, new ClusterStateUpdateSchema(node.getVersion(), true));
    }

public static BytesReference serializeClusterState(Diff<ClusterState> clusterStateDiff, DiscoveryNode node) throws IOException {
        return serializeClusterStateOrDiff(clusterStateDiff, new ClusterStateUpdateSchema(node.getVersion(), false));
    }

public static BytesReference serializeClusterStateForJoin(Writeable writer, DiscoveryNode node)
        throws IOException {
        return serializeClusterStateOrDiff(writer, new NodeVersionSchemaBase(node.getVersion()));
    }
```

And finally the end method will look something like 

```
private static BytesReference serializeClusterStateOrDiff(Writeable writer,
                                                              NodeVersionSchemaBase streamHeaders) throws IOException {
        final BytesStreamOutput bStream = new BytesStreamOutput();
        try (StreamOutput stream = new OutputStreamStreamOutput(CompressorFactory.COMPRESSOR.threadLocalOutputStream(bStream))) {
            streamHeaders.writeTo(stream);
            writer.writeTo(stream);
        }
        final BytesReference serializedByteRef = bStream.bytes();
        logger.trace("serialized writable object for node version [{}] with size [{}]", streamHeaders.getNodeVersion(),
            serializedByteRef.length());
        return serializedByteRef;
    }
```

Again, here I'm using Util class itself to create these schema objects as currently we've already moved that complexity from JoinHelper as well as PTH. If needed this can be directly created from both of those 2 classes, as they are the callers and it can be their responsibility. Right now, it's not a big schema, so keeping it here is also fine from my side. As long as we keep separate methods for separate callers like I created a new method for Node Join.

Extendibility support :

2nd approach allows us to modify the headers for one caller without touching anything for second caller and common method doesn't need any changes. Only schema classes would require changes, and schema creation code.

Please suggest if 2nd approach looks good, or what more improvements can be done there ?
**Note 1** : In case JoinHelper needs to add a different variable in stream, we'll create a new class similar to `ClusterStateUpdateSchema` may be `ValidateJoinStateSchema` and add data there. This NodeVersionBaseSchema is added so it's always added to the stream (else serialization fails for ClusterState).
**Note 2**: I thought of using composition here also instead of inheritance, but wanted to keep the end method with same type(NodeVersionSchemaBase streamHeaders, so any caller can pass their headers as they all will be inherited from base class) for stream headers, so used inheritance to avoid any if else (or more methods).

Reply:
I was thinking of something a little simpler. Essentially replace `ClusterStateUtils` with this class:

```java
public final class CompressedStreamUtils {
    public static BytesReference createCompressedStream(Version version, CheckedConsumer<StreamOutput, IOException> outputConsumer) throws IOException {
        final BytesStreamOutput bStream = new BytesStreamOutput();
        try (StreamOutput stream = new OutputStreamStreamOutput(CompressorFactory.COMPRESSOR.threadLocalOutputStream(bStream))) {
            stream.setVersion(version);
            outputConsumer.accept(stream);
        }
        return bStream.bytes();
    }

    public static StreamInput decompressBytes(BytesTransportRequest request, NamedWriteableRegistry namedWriteableRegistry)
        throws IOException {
        final Compressor compressor = CompressorFactory.compressor(request.bytes());
        final StreamInput in;
        if (compressor != null) {
            in = new InputStreamStreamInput(compressor.threadLocalInputStream(request.bytes().streamInput()));
        } else {
            in = request.bytes().streamInput();
        }
        in.setVersion(request.version());
        return new NamedWriteableAwareStreamInput(in, namedWriteableRegistry);
    }
}
```

Both PTH and JoinHelper can use those functions and retain full control over what they read and write from the stream. It seems like this should work? I did a [quick prototype here](https://github.com/andrross/OpenSearch/commit/ea757e21d665871afa33ccc7fabd67cdfe2cee4a) (didn't update the tests though).

Reply:
@andrross  

Sure I saw the prototype code. Only difference I see is that we're writing boolean from PTH class itself. 

I suggested this also in my previous [comment](https://github.com/opensearch-project/OpenSearch/pull/7321#discussion_r1214268115) 

>Again, here I'm using Util class itself to create these schema objects as currently we've already moved that complexity from JoinHelper as well as PTH. If needed this can be directly created from both of those 2 classes, as they are the callers and it can be their responsibility.

So we can create those - `new ClusterStateUpdateSchema(node.getVersion(), false)` from PTH class and pass it in the util method  -> 

`public static BytesReference serializeClusterStateOrDiff(Writeable writer, NodeVersionSchemaBase streamHeaders)`

I hope that is similar to what you're suggesting :) . I'm okay if you want that in `CheckedConsumer<StreamOutput, IOException> outputConsumer` format.

@shwetathareja  @andrross  
Let's finalize on one coding approach. Thanks Andrew for the prototype code.

@shwetathareja
Let me know what do you prefer. I'll make code changes accordingly.

Reply:
@amkhar Lets keep the code simple and let the callers PTH and JoinHelper pass the supplier to handle the serialization of their objects (with/ without boolean). We don't need new interface and classes.

Reply:
Thank you both.

@andrross I am planning to take your code and push a new commit out of it.

Please explain if there is any other change which is required after this ?

I want to incorporate all required changes in one go.

Reply:
No other comments from my side. Thanks!

Reply:
Thanks @andrross .